src/storage/src/table/state_table.rs

```diff
@@ -113,171 +116,109 @@ impl<S: StateStore> StateTable<S> {
-    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
-
-    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
-    /// discarded.
-    cell_based_item: Option<(Vec<u8>, Row)>,
-    pk_serializer: OrderedRowSerializer,
+struct StateTableRowIter<S: StateStore> {
+    _phamtom: PhantomData<S>,
 }
-type MemTableIter<'a> = btree_map::Iter<'a, Row, RowOp>;
 
-enum NextOutcome {
-    MemTable,
-    Storage,
-    Both,
-    End,
-}
-impl<'a, S: StateStore> StateTableRowIter<'a, S> {
-    async fn new(
-        keyspace: &Keyspace<S>,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`.
+impl<S: StateStore> StateTableRowIter<S> {
+    /// This function scans kv pairs from the `shared_storage`(`cell_based_table`) and
+    /// memory(`mem_table`). If a record exist in both `cell_based_table` and `mem_table`, result
+    /// `mem_table` is returned according to the operation(RowOp) on it.
+    #[try_stream(ok = Option<Row>, error = StorageError)]
+    async fn into_stream<'a>(
+        keyspace: &'a Keyspace<S>,
         table_descs: Vec<ColumnDesc>,
-        mem_table_iter: Peekable<MemTableIter<'a>>,
-        order_types_vec: &[OrderType],
+        mut mem_table_iter: MemTableIter<'a>,
+        order_types_vec: &'a [OrderType],
         epoch: u64,
-    ) -> StorageResult<StateTableRowIter<'a, S>> {
-        let mut cell_based_streaming_iter =
+    ) {
+        let mut cell_based_table_iter =
             CellBasedTableStreamingIter::new(keyspace, table_descs, epoch).await?;
         let pk_serializer = OrderedRowSerializer::new(order_types_vec.to_vec());
-        let cell_based_item = cell_based_streaming_iter.next().await.map_err(err)?;
-        let state_table_iter = Self {
-            mem_table_iter,
-            cell_based_streaming_iter,
-            cell_based_item,
-            pk_serializer,
-        };
-
-        Ok(state_table_iter)
-    }
 
-    /// This function scans kv pairs from the `shared_storage`(`cell_based_table`) and
-    /// memory(`mem_table`). If a record exist in both `cell_based_table` and `mem_table`, result in
-    /// `mem_table` is returned according to the operation(RowOp) on it. This is because the data in
-    /// memory is fresher.
-    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let mut cell_based_table_item = cell_based_table_iter.next().await.map_err(err)?;
+        let mut mem_table_item = mem_table_iter.next();
         loop {
-            let next_flag;
-            let res;
-            let cell_based_item = self.cell_based_item.take();
-            match (cell_based_item, self.mem_table_iter.peek()) {
+            match (&cell_based_table_item, &mem_table_item) {
                 (None, None) => {
-                    next_flag = NextOutcome::End;
-                    res = None;
+                    yield None;
```

Comment:
Does `yield None` mean that the stream has ended? If so, we can simply `break` here so that the stream is finished, and the type for this stream could also be `impl Stream<Item = StorageResult<Row>>`.

Reply:
If `break`, the caller would need to change to match an error.

Maybe put in your `cow` pr. 🤣